module13and14/LoopCanvas/LoopCanvas/Views/CanvasViewModel.swift

```diff
@@ -0,0 +1,345 @@
+  }
+
+  func updateBlockDragLocation(block: Block, location: CGPoint) {
+    if !block.dragging {
+      startBlockDrag(block: block)
+    }
+    block.location = location
+  }
+
+  func startBlockDrag(block: Block) {
+    block.dragging = true
+    if let blockGroup = block.blockGroup {
+      canvasModel.removeBlockFromBlockGroup(block: block, blockGroup: blockGroup)
+    }
+    if !block.isLibraryBlock {
+      draggingBlock = block
+    }
+    updateAllBlocksList()
+  }
+
+  func dropBlockOnCanvas(block: Block) -> Block {
+    // TODO - break this function up and refator logic into Canvas Model
+
+    block.dragging = false
+    draggingBlock = nil
+
+    var blockDroppedOnCanvas = block
+    if block.isLibraryBlock {
+      blockDroppedOnCanvas = Block(
+        id: Block.getNextBlockId(),
+        location: CGPoint(x: block.location.x - canvasScrollOffset.x, y: block.location.y - canvasScrollOffset.y),
+        color: block.color,
+        icon: block.icon,
+        visible: true,
+        loopURL: block.loopURL,
+        relativePath: block.relativePath
+      )
+
+      syncBlockLocationsWithSlots() // reset library block location
+    }
+
+    let blockAddedToGroup = canvasModel.checkBlockPositionAndAddToAvailableGroup(block: blockDroppedOnCanvas)
+
+    if !blockAddedToGroup {
+      if blockDroppedOnCanvas.location.y > canvasModel.library.libaryFrame.minY + CanvasViewModel.blockSize / 2 {
+        // If the block is re-dropped on the library, delete it.
+        // Right now we dont need to do anything as the block is
+        // not a member of a group and has been removed from the library,
+        // and draggingBlock = nil so the block should simply disappear.
+      } else {
+        canvasModel.addBlockGroup(initialBlock: blockDroppedOnCanvas)
+      }
+    }
+
+    updateAllBlocksList()
+
+    return blockDroppedOnCanvas
+  }
+
+  func selectLoopCategory(categoryName: String) {
+    canvasModel.library.setLoopCategory(categoryName: categoryName)
+    updateAllBlocksList()
+    syncBlockLocationsWithSlots()
+  }
+}
+
+// Canvas managmeent events
+
+extension CanvasViewModel {
+  func clearCanvas() {
+    canvasModel.clear()
+    updateAllBlocksList()
+  }
+
+  func renameSong(newName: String, thunbnail: UIImage?) {
+    canvasModel.name = newName
+    canvasModel.thumnail = thunbnail
+
+    saveSong()
+  }
+
+  func saveSong() {
+    guard let documentsDirectory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
+      print("Error: Unable to access documents directory")
+      return
+    }
+    let dataDirectoryURL = documentsDirectory.appendingPathComponent("LoopCanvas")
+    if !FileManager.default.fileExists(atPath: dataDirectoryURL.path) {
+      do {
+        try FileManager.default.createDirectory(at: dataDirectoryURL, withIntermediateDirectories: true, attributes: nil)
+      } catch {
+        print("Error: Unable to create directory \(dataDirectoryURL)")
```

Comment:
Don't use `print` explore using the `Logger` clases and the `OSLog` library. It will send your statements to the console just like print, but, you can set it not to pollute your user's consoles when the app is live. You can do a bunch of other things as well. AND, unlike print it runs on background threads correctly. `print` does not and sometimes causes wonkiness with threaded code.

Reply:
Yes good suggestion. Done.